### PR TITLE
Convert ReactMeteorData to a Higher-Order Component

### LINKED ISSUE #200: Remove "react-addons-pure-render-mixin"?
I have a question. In regards to this post, [https://facebook.github.io/react/blog/2016/07/13/mixins-considered-harmful.html#solution](https://facebook.github.io/react/blog/2016/07/13/mixins-considered-harmful.html#solution) would it be possible to remove "react-addons-pure-render-mixin" from "react-meteor-data"? They suggest to start using `var shallowCompare = require('react-addons-shallow-compare');` for `shouldComponentUpdate` I am not sure if this is possible or not? Thanks

### LINKED ISSUE #216: Updating react-addons-pure-render-mixin to React.PureComponent
React 15.5 is discontinuing support for React add-ons, including pure-render-mixin, as mentioned in their release notes: https://facebook.github.io/react/blog/2017/04/07/react-v15.5.0.html

Plans to update meteor-react-data?

### PR DESCRIPTION
**Pull request by leocavalcante**
Converts the ReactMeteorData mixin to a [Higher-Order Component](https://facebook.github.io/react/docs/higher-order-components.html)

This PR fixes warnings with latest React like:
* React.createClass is deprecated and will be removed in version 16.

Fixes #216
Closes #200 
(maybe more)

### CODE REVIEW

File: packages/react-meteor-data/ReactMeteorData.jsx [hunk 1]
```diff
@@ -1,36 +1,52 @@
...
-      this.props = nextProps;
-      this.state = nextState;
-      newData = this._meteorDataManager.calculateData();
-    } finally {
-      this.props = saveProps;
-      this.state = saveState;
+import React from 'react';
+
+export function connect({ getMeteorData }, WrappedComponent) {
+  class ReactMeteorData extends React.Component {
```
Comment (s-ol) [inline packages/react-meteor-data/ReactMeteorData.jsx]: shouldn't this be a `React.PureComponent`?

Comment (menelike) [inline packages/react-meteor-data/ReactMeteorData.jsx]: ... and sometimes it should not be pure at all, this is why this is optional and set to true per default https://github.com/meteor/react-packages/blob/2e0cbe94b1a8d657559e223006ed0507b212829c/packages/react-meteor-data/createContainer.jsx#L20

Comment (leocavalcante) [inline packages/react-meteor-data/ReactMeteorData.jsx]: Working on it.

**Comment by leocavalcante** [comment]
My test app, since there is no unit tests, was:

```js
import React, { Component } from 'react'
import { render } from 'react-dom'
import { Meteor } from 'meteor/meteor'
import { Mongo } from 'meteor/mongo'
import { createContainer } from 'meteor/react-meteor-data'

class MessageList extends Component {
  render() {
    const { messages } = this.props

    console.log(messages)

    return (
      <ul>
        {messages.map(message => <li>{message.body}</li>)}
      </ul>
    )
  }
}

const Messages = new Mongo.Collection('messages')

if (Meteor.isServer) {
  Meteor.publish('messages', () => {
    return Messages.find({})
  })
}

if (Meteor.isClient) {
  const Container = createContainer(() => {
    const handle = Meteor.subscribe('messages')
    const messages = Messages.find({}).fetch()

    return {
      ready: handle.ready(),
      messages
    }
  }, MessageList)

  Meteor.startup(() => render(<Container />, document.getElementById('root')))
}
```

And I always get impressed how easy is to build a real-time database-backed app with Meteor :open_mouth:

File: packages/react-meteor-data/ReactMeteorData.jsx [hunk 2]
```diff
@@ -1,36 +1,54 @@
...
-      // See https://github.com/facebook/react/issues/3398.
-      this.props = nextProps;
-      this.state = nextState;
-      newData = this._meteorDataManager.calculateData();
-    } finally {
-      this.props = saveProps;
-      this.state = saveState;
+import React from 'react';
+
+export function connect({ getMeteorData, pure = true }, WrappedComponent) {
```
Comment (leocavalcante) [inline packages/react-meteor-data/ReactMeteorData.jsx]: Hi @benjamn, thanks for review. There is no particular reason. Actually, the [convention for `connect`](https://facebook.github.io/react/docs/higher-order-components.html#convention-maximizing-composability) is to be a high-order function that returns a high-component. Like Redux's `connect` and Apollo's `graphql`. You think we should go by that?

Comment (dburles) [inline packages/react-meteor-data/ReactMeteorData.jsx]: Related https://github.com/meteor/react-packages/pull/215

**Comment by hwillson** [comment]
Hi @leocavalcante - would you mind rebasing this PR? As a temporary fix we just merged PR #218, to help get rid of the `React.createClass` deprecation warnings. This PR represents a better long term solution however, so if you're able to rebase, I'll start doing some thorough testing of these changes. Thanks!

**Comment by leocavalcante** [comment]
Hi @hwillson - here is it.

File: packages/react-meteor-data/react-meteor-data.jsx [hunk 3]
```diff
@@ -1,11 +1,9 @@
...
   react: '15.x',
-  'react-addons-pure-render-mixin': '15.x',
-  'create-react-class': '15.x',
 }, 'react-meteor-data');
 
 const createContainer = require('./createContainer.jsx').default;
-const ReactMeteorData = require('./ReactMeteorData.jsx').default;
 
-export { createContainer, ReactMeteorData };
+export { createContainer };
```
Comment (benjamn) [inline packages/react-meteor-data/react-meteor-data.jsx]: You should be able to do the remaining export on one line:
```js
export { default as createContainer } from './createContainer.jsx';
```

File: packages/react-meteor-data/react-meteor-data.jsx [hunk 4]
```diff
@@ -1,11 +1,9 @@
 import { checkNpmVersions } from 'meteor/tmeasday:check-npm-versions';
+
 checkNpmVersions({
   react: '15.x',
-  'react-addons-pure-render-mixin': '15.x',
-  'create-react-class': '15.x',
```
Comment (benjamn) [inline packages/react-meteor-data/react-meteor-data.jsx]: Why not keep checking `create-react-class` here?

Comment (leocavalcante) [inline packages/react-meteor-data/react-meteor-data.jsx]: It is not used anymore.

**Comment by fabien-h** [comment]
When we `meteor add ...` We still have the 0.2.9 and not the 0.2.10.

    meteor add react-meteor-data@0.2.10
    error: no such version react-meteor-data@0.2.10

**Comment by hwillson** [comment]
Hi @fabien-h - version 0.2.10 hasn't been published yet. The most recent published version is [0.2.9](https://atmospherejs.com/meteor/react-meteor-data). A new (published) version is coming soon though - thanks!

**Comment by abernix** [comment]
If anyone wants to try this right now, I've published `react-meteor-data@0.2.11-beta.1` which includes this, on top of #218 (which is essentially nullified, but worth saving as a worth contribution by @merlinpatt).

Feedback appreciated!

**Comment by rijk** [comment]
I have a weird issue with this. Changes in context don't seem to propagate through this component.. The context is there, but an update doesn't come through when it has to pass down the tree through a `ReactMeteorData` container. They do through normal components.

**Comment by rijk** [comment]
Figured it out, it works when I use the `pure: false` option.

**Comment by mhagmajer** [comment]
Doesn't this break the default export of https://github.com/meteor/react-packages/blob/devel/packages/react-meteor-data/ReactMeteorData.jsx ?

In our project we're using this in the following way:

```javascript
import { ReactMeteorData } from 'meteor/react-meteor-data';
import React from 'react';

export default class MeteorComponent<D, P, S, M> extends React.Component<D, P, S> {
  static defaultProps: D;
  props: P;
  state: S;
  data: M;

  getMeteorData(): M {
    return ({}: any);
  }

  componentWillMount() {
    ReactMeteorData.componentWillMount.bind(this)();
  }

  componentWillUpdate(nextProps: P, nextState: S) {
    ReactMeteorData.componentWillUpdate.bind(this)(nextProps, nextState);
  }

  componentWillUnmount() {
    ReactMeteorData.componentWillUnmount.bind(this)();
  }
}
```

And `MeteorComponent` is the base class for all our react component that can track changes to Meteor data.

**Comment by leocavalcante** [comment]
You are right. Sorry about that. This change is actually a BC breaker since `ReactMeteorData` isn't a plain object anymore and the package didn't export it.

As workaround, you should be able to get those methods back as a dummy container prototype:
```js
import { createContainer } from 'meteor/react-meteor-data'

const MeteorContainer = createContainer(() => { }, () => { })
const ReactMeteorData = MeteorContainer.prototype

console.log(ReactMeteorData.componentWillMount)
```

To @benjamn @hwillson and @abernix 
I'm not in the position to make a call here, but I'll be glad to make a PR to fix whatever you decide 😃

**Comment by hwillson** [comment]
Thanks for the heads up @mhagmajer. We've gone ahead and published the new changes since they were being asked for by a large number of users. As you've mentioned though, the new version does break backwards compatibility in some cases. To make sure we don't lose track of this, would you mind creating a new issue for this, outlining exactly what you've commented here? We can then continue the discussion in that new issue thread, and reach out to others for feedback (as well as gather 👍's to see how many people are affected by this). Sorry for any inconvenience this has caused. For now you can consider version pinning at 0.2.9. Thanks!

**Comment by hwillson** [comment]
@mhagmajer PR #224 should help resolve your issue. Hopefully we can get it merged/published soon. Thanks!

**Pull request merged**